CMake: fix missing `SWIFT_CONCURRENCY_GLOBAL_EXECUTOR` value

MaxDesiatov opened the pull request:
`SWIFT_CONCURRENCY_GLOBAL_EXECUTOR` is defined in `stdlib/cmake/modules/StdlibOptions.cmake`, which is not included during the first pass of evaluation of the root `CMakeLists.txt`. It is available on subsequent evaluations after the value is stored in CMake cache. This led to subtle bugs, where `usr/lib/swift_static/linux/static-stdlib-args.lnk` didn't contain certain flags on clean toolchain builds, but did contain them in incremental builds.

Not having these autolinking flags in toolchain builds leads to errors when statically linking executables on Linux.

Additionally, since are trivial tests previously didn't link Dispatch statically, the didn't expose a bug where `%import-static-libdispatch` substitution had a missing value. To fix that I had to update `lit.cfg` and clean up some of the related path computations to infer a correct substitution value.

Resolves some of the errors reported in https://github.com/apple/swift/issues/65097.

MaxDesiatov committed:
CMake: fix missing `SWIFT_CONCURRENCY_GLOBAL_EXECUTOR` value

`SWIFT_CONCURRENCY_GLOBAL_EXECUTOR` is defined in `stdlib/cmake/modules/StdlibOptions.cmake`, which is not included during the first pass of evaluation of the root `CMakeLists.txt`. It is available on subsequent evaluations after the value is stored in CMake cache. This led to subtle bugs, where `usr/lib/swift_static/linux/static-stdlib-args.lnk` didn't contain certain flags on clean toolchain builds, but did contain them in incremental builds.

Not having these autolinking flags in toolchain builds leads to errors when statically linking executables on Linux.

MaxDesiatov commented:
@swift-ci test

MaxDesiatov commented:
@swift-ci build toolchain

MaxDesiatov commented:
Running a first pass of tests before running cross-repo integration tests...

al45tair commented:
> Potentially resolves #65097.

Isn't this going to auto-close that issue? So either it _does_ resolve it, or we should adjust the message wording here?

MaxDesiatov commented:
I'm not merging this one until we have appropriate integration tests enabled, will update the wording after that.

MaxDesiatov committed:
Fix `-DSWIFT_PATH_TO_LIBDISPATCH_STATIC_BUILD` value

MaxDesiatov commented:
@swift-ci test

MaxDesiatov commented:
@swift-ci build toolchain

MaxDesiatov commented on test/lit.cfg:
```diff
@@ -1593,19 +1593,13 @@ elif (run_os in ['linux-gnu', 'linux-gnueabihf', 'freebsd', 'openbsd', 'windows-
+    libdispatch_static_artifact_dir = os.path.join(config.libdispatch_static_build_path, 'lib')
     libdispatch_static_artifacts = [
-        make_path(libdispatch_static_artifact_dir, 'src', 'libdispatch.a'),
-        make_path(libdispatch_static_artifact_dir, 'src', 'swift', 'libswiftDispatch.a'),
-        make_path(libdispatch_swift_static_module_dir, 'Dispatch.swiftmodule')]
```
We don't need a Swift overlay for Dispatch to statically linked a simple executable, Dispatch is used directly without an overlay from the concurrency runtime.

MaxDesiatov commented on test/lit.cfg:
```diff
@@ -1593,19 +1593,13 @@ elif (run_os in ['linux-gnu', 'linux-gnueabihf', 'freebsd', 'openbsd', 'windows-
             % (libdispatch_source_dir, libdispatch_swift_module_dir, libdispatch_artifact_dir))
 
-    libdispatch_static_artifact_dir = config.libdispatch_static_build_path
-    libdispatch_swift_static_module_dir = make_path(libdispatch_static_artifact_dir, 'src', 'swift', 'swift')
```
Overlay path deleted as unused.

MaxDesiatov commented on test/lit.cfg:
```diff
@@ -1593,19 +1593,13 @@ elif (run_os in ['linux-gnu', 'linux-gnueabihf', 'freebsd', 'openbsd', 'windows-
 
-    libdispatch_static_artifact_dir = config.libdispatch_static_build_path
-    libdispatch_swift_static_module_dir = make_path(libdispatch_static_artifact_dir, 'src', 'swift', 'swift')
+    libdispatch_static_artifact_dir = os.path.join(config.libdispatch_static_build_path, 'lib')
```
Added a `lib` component for easier reuse of this variable in subsequent expressions below this line.

MaxDesiatov commented on test/lit.cfg:
```diff
@@ -1593,19 +1593,13 @@ elif (run_os in ['linux-gnu', 'linux-gnueabihf', 'freebsd', 'openbsd', 'windows-
-    libdispatch_swift_static_module_dir = make_path(libdispatch_static_artifact_dir, 'src', 'swift', 'swift')
+    libdispatch_static_artifact_dir = os.path.join(config.libdispatch_static_build_path, 'lib')
     libdispatch_static_artifacts = [
-        make_path(libdispatch_static_artifact_dir, 'src', 'libdispatch.a'),
```
Built artifacts are located in `lib`, not `src`.

MaxDesiatov commented on test/lit.cfg:
```diff
@@ -1593,19 +1593,13 @@ elif (run_os in ['linux-gnu', 'linux-gnueabihf', 'freebsd', 'openbsd', 'windows-
-            make_path(libdispatch_static_artifact_dir, 'src'),
-            make_path(libdispatch_static_artifact_dir, 'src', 'BlocksRuntime'),
-            make_path(libdispatch_static_artifact_dir, 'src', 'swift')))
+        config.import_libdispatch_static = '-L %s' % libdispatch_static_artifact_dir
```
Since neither the overlay nor the headers are used for static linking, this is simplified to a single library search path flag.

MaxDesiatov commented on test/lit.cfg:
```diff
@@ -2649,6 +2643,10 @@ run_filecheck = '%s %s --allow-unused-prefixes --sanitize BUILD_DIR=%s --sanitiz
+# a pure string substitution without understanding that these components are grouped together. That is, the following
+# subsitution name can't be `%import-libdispatch-static`, otherwise the first two components will be substituted with
+# the value of `%import-libdispatch` substitution with `-static` string appended to it.
 config.substitutions.append(('%import-static-libdispatch', getattr(config, 'import_libdispatch_static', '')))
```
While working on this fix I attempted to rename this to `%import-libdispatch-static` for consistency, which led to a hard to debug test failure due to incorrect flags passed. Added a note about this substitution behavior for posterity.

MaxDesiatov commented on utils/build-script-impl:
```diff
@@ -1901,7 +1901,7 @@ for host in "${ALL_HOSTS[@]}"; do
                     -DSWIFT_PATH_TO_LIBDISPATCH_SOURCE:PATH="${LIBDISPATCH_SOURCE_DIR}"
                     -DSWIFT_PATH_TO_LIBDISPATCH_BUILD:PATH="$(build_directory ${host} libdispatch)"
-                    -DSWIFT_PATH_TO_LIBDISPATCH_STATIC_BUILD:PATH="$(build_directory ${host} libdispatch_static)"
+                    -DSWIFT_PATH_TO_LIBDISPATCH_STATIC_BUILD:PATH="$(build_directory ${host} swift)/$(basename $(build_directory ${host} libdispatch))-static-prefix"
```
Dispatch built as a dependency of the Swift compiler is located in the Swift build subdirectory, not in its own directory at the top level of the build.

etcwilde replied:
IIRC, this is the dispatch built for the concurrency runtime. I think this will work for this use-case, but it is probably worth noting.

MaxDesiatov replied:
For concurrency runtime and SourceKIT IIUC

MaxDesiatov commented:
https://github.com/apple/swift-integration-tests/pull/78

@swift-ci build toolchain linux

MaxDesiatov commented:
@swift-ci build toolchain

tomerd commented:
seems reasonable to me. should any of this be separated into a dedicated "cleanup" PR, or do they all need to be done together?

tomerd commented:
I meant the other order, ie first do the cleanup then fix the static linking initialization issue. but not a big deal either way

MaxDesiatov commented:
AFAIU this repository prefers merge commits, so commit separation will be preserved after the PR is merged. If having separate PRs would make the review easier I can split it, but the diff seems relatively small to me from that perspective.

drexin commented on CMakeLists.txt:
```diff
@@ -202,6 +202,10 @@ set(SWIFT_HOST_VARIANT_ARCH "${SWIFT_HOST_VARIANT_ARCH_default}" CACHE STRING
 
+# Subsequent options may refer to `StdlibOptions`, which have to be defined first.
+list(APPEND CMAKE_MODULE_PATH ${CMAKE_CURRENT_LIST_DIR}/stdlib/cmake/modules)
+include(StdlibOptions)
```
Instead of including `StdlibOptions` here, why don't we move the following into `StdlibOptions`?

```
# Use dispatch as the system scheduler by default.
# For convenience, we set this to false when concurrency is disabled.
set(SWIFT_CONCURRENCY_USES_DISPATCH FALSE)
if(SWIFT_ENABLE_EXPERIMENTAL_CONCURRENCY AND "${SWIFT_CONCURRENCY_GLOBAL_EXECUTOR}" STREQUAL "dispatch")
  set(SWIFT_CONCURRENCY_USES_DISPATCH TRUE)
endif()
```

AFAICT it is only used in the stdlib build, so really has no business being set in the root `CMakeLists.txt`

etcwilde replied:
^ Was just thinking that.

MaxDesiatov replied:
Good point, that's addressed now.

MaxDesiatov committed:
Move `SWIFT_CONCURRENCY_USES_DISPATCH` to `StdlibOptions.cmake`

It is only used in the stdlib build, so really has no business being set in the root `CMakeLists.txt`.

MaxDesiatov committed:
Remove redundant `include(StdlibOptions)`

MaxDesiatov commented on CMakeLists.txt:
```diff
@@ -687,13 +687,6 @@ if(NOT EXISTS "${SWIFT_PATH_TO_SWIFT_SYNTAX_SOURCE}")
-set(SWIFT_CONCURRENCY_USES_DISPATCH FALSE)
-if(SWIFT_ENABLE_EXPERIMENTAL_CONCURRENCY AND "${SWIFT_CONCURRENCY_GLOBAL_EXECUTOR}" STREQUAL "dispatch")
-  set(SWIFT_CONCURRENCY_USES_DISPATCH TRUE)
-endif()
```
Moved to `StdlibOptions.cmake`

MaxDesiatov commented:
@swift-ci test

etcwilde reviewed (approved):
This is a fair bit better.

The sourcekit should be a misnomer coming from our build-soup (don't have to fix it now). We're not building dispatch for the "host", we're building it for the "target", so it may not work in a cross-compiling situation.

MaxDesiatov commented:
@swift-ci build toolchain

MaxDesiatov commented:
@swift-ci test linux
